pyquil/reference_simulator.py

```diff
@@ -163,8 +177,37 @@ class ReferenceDensitySimulator(AbstractQuantumSimulator):
     def __init__(self, n_qubits: int, rs: RandomState = None):
         self.n_qubits = n_qubits
         self.rs = rs
-        self.density = np.zeros((2 ** n_qubits, 2 ** n_qubits), dtype=np.complex128)
-        self.density[0, 0] = complex(1.0, 0)
+        self.initial_density = None
+        # set self.density via
+        self.reset()
+
+    def set_initial_state(self, state_matrix):
+        """
+        This method is the correct way (TM) to set the state matrix to a state other than the
+        default state. The default state of ReferenceDensitySimulator is |000...00>.
+
+        To restore default behavior and state of ReferenceDensitySimulator use
+        ``state_matrix = None.``
+
+        :param state_matrix: numpy.ndarray or None.
+        :return: ``self`` to support method chaining.
+        """
+        if state_matrix is None:
+            self.initial_density = None
+        else:
+            rows, cols = state_matrix.shape
+            if rows != cols:
+                raise ValueError("The state matrix is not square.")
+            if self.n_qubits != int(np.log2(rows)):
+                raise ValueError("The state matrix is not defined on the same numbers of qubits as "
+                                 "the QVM.")
+            if _is_valid_quantum_state(state_matrix):
+                self.initial_density = state_matrix
+            else:
+                raise ValueError("The state matrix is not valid. It must be Hermitian, trace one, "
+                                 "and have non-negative eigenvalues.")
+        self.reset()
```

Comment:
As per private discussion, don't reset. And make the docstring reflect that this does *not* mutate the current state.

Reply:
- [ ] todo